sdks/java/harness/src/main/java/org/apache/beam/fn/harness/PrecombineGroupingTable.java

```diff
@@ -17,434 +17,283 @@
-  @Override
-  public void put(Object pair, Receiver receiver) throws Exception {
-    put(
-        (K) pairInfo.getKeyFromInputPair(pair),
-        (InputT) pairInfo.getValueFromInputPair(pair),
-        receiver);
+    public void add(InputT value) {
+      dirty = true;
+      accumulator =
+          combineFn.addInput(
+              accumulator, value, options, NullSideInputReader.empty(), groupingKey.getWindows());
+      accumulatorSize = accumulatorSizer.estimateSize(accumulator);
+    }
   }
 
   /**
    * Adds the key and value to this table, possibly flushing some entries to output if the table is
    * full.
    */
-  public void put(K key, InputT value, Receiver receiver) throws Exception {
-    Object groupingKey = groupingKeyCreator.createGroupingKey(key);
-    GroupingTableEntry<K, InputT, AccumT> entry = table.get(groupingKey);
-    if (entry == null) {
-      entry = createTableEntry(key);
-      table.put(groupingKey, entry);
-      size += PER_KEY_OVERHEAD;
-    } else {
-      size -= entry.getSize();
-    }
-    entry.add(value);
-    size += entry.getSize();
-
-    if (size >= maxSize) {
-      long targetSize = (long) (TARGET_LOAD * maxSize);
-      Iterator<GroupingTableEntry<K, InputT, AccumT>> entries = table.values().iterator();
-      while (size >= targetSize) {
-        if (!entries.hasNext()) {
-          // Should never happen, but sizes may be estimates...
-          size = 0;
-          break;
+  @VisibleForTesting
+  public void put(
+      WindowedValue<KV<K, InputT>> value, FnDataReceiver<WindowedValue<KV<K, AccumT>>> receiver)
+      throws Exception {
+    // Ignore timestamp for grouping purposes.
+    // The Pre-combine output will inherit the timestamp of one of its inputs.
+    WindowedValue<Object> groupingKey =
+        WindowedValue.of(
+            keyCoder.structuralValue(value.getValue().getKey()),
+            IGNORED,
+            value.getWindows(),
+            value.getPane());
+
+    GroupingTableEntry entry =
+        lruMap.compute(
+            groupingKey,
+            (key, tableEntry) -> {
+              if (tableEntry == null) {
+                tableEntry =
+                    new GroupingTableEntry(
+                        groupingKey, value.getValue().getKey(), value.getValue().getValue());
+              } else {
+                tableEntry.add(value.getValue().getValue());
+              }
+              return tableEntry;
+            });
+    weight += entry.getWeight();
+    // Increase the maximum only if we require it
+    maxWeight.accumulateAndGet(weight, (current, update) -> current < update ? update : current);
+
+    // Update the cache to ensure that LRU is handled appropriately and for the cache to have an
+    // opportunity to shrink the maxWeight if necessary.
+    cache.put(Key.INSTANCE, this);
+
+    // Get the updated weight now that the cache may have been shrunk and respect it
+    long currentMax = maxWeight.get();
+    if (weight > currentMax) {
```

Comment:
If this is triggered by shrink() why not do it in shrink but instead rely on new input?